File: 5-longest-palindromic-substring/5-longest-palindromic-substring.cpp
class Solution {
public:
    string longestPalindrome(string s) {
    bool dp[s.length()][s.length()];
    int a = -1;
    int b = -1;
    for (int diagonal = 0; diagonal < s.length(); diagonal++) {
        for (int row = 0, col = diagonal; col < s.length(); row++, col++) {
            if (diagonal == 0) {
                dp[row][col] = true;
            } else if (diagonal == 1) {
                if (s[row] == s[col]) {
                    dp[row][col] = true;
                } else {
                    dp[row][col] = false;
                }
            } else {
                if (s[row] == s[col] && dp[row + 1][col - 1]) {
                    dp[row][col] = true;
                } else {
                    dp[row][col] = false;
                }
            }
            if (dp[row][col]) {
                a = row;
                b = diagonal + 1;
            }
        }
    }
    return s.substr(a,b);
}
};

File: 5-longest-palindromic-substring/5-longest-palindromic-substring.java
class Solution {
    public String longestPalindrome(String s) {
        int n=s.length();
        int[][] dp = new int[n][n];
        String res="";
        

        for(int gap=0;gap<n;gap++){
            for(int i=0,j=gap;j<s.length();i++,j++){
                if(gap==0){
//                     if it is single character means its a palindrome,store true in dp.
                    dp[i][j]=1;
                }
                else if(gap==1){
//                     if it is two characters means its a palindrome,store true in dp.
                    if(s.charAt(i)==s.charAt(j)) dp[i][j]=1;
                }
                else{
//                    compare the boundary char's and if inner subString(i+1,j-1) is palindrome the whole string is palindrome.

                    if(s.charAt(i)==s.charAt(j) && dp[i+1][j-1]==1) dp[i][j]=1;
                }
//                 each time the gap is constant ,at specific gap if we found palindrome then its current longest palindrome.
//                 subsequently on each gaps we find the longest palindrome.
                if(dp[i][j]==1) res = s.substring(i,j+1);
            }
        }
        return res;
    }
}